### PR TITLE
add prometheus-mongodb-exporter

### PR DESCRIPTION
**Pull request by undefined-moe**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix [hunk 1]
```diff
@@ -0,0 +1,71 @@
+{ config, lib, pkgs, options }:
+
+with lib;
+
+let
+  cfg = config.services.prometheus.exporters.mongodb;
+in
+{
+  port = 9100;
```
Comment (NickCao) [inline nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix:9]: This seems unused.

File: nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  extraOpts = {
+    uri = mkOption {
+      type = types.str;
+      default = "mongodb://localhost:27017/test";
```
Comment (NickCao) [inline nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix:13]: test?

File: nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      example = "mongodb://localhost:27017/test";
+      description = lib.mdDoc ''
+        MongoDB URI to connect to.
+      '';
+    };
+    collStats = mkOption {
+      type = types.listOf types.str;
+      default = [ "test" ];
```
Comment (NickCao) [inline nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix:21]: Just leave the default empty list.

File: nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      example = [ "db1.coll1" "db2" ];
+      description = lib.mdDoc ''
+        List of comma separared databases.collections to get $collStats
+      '';
+    };
+    indexStats = mkOption {
+      type = types.listOf types.str;
+      default = [ "test" ];
```
Comment (NickCao) [inline nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix:29]: Same here.

File: nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    telemetryPath = {
+      type = types.str;
+      default = "/metrics";
+      example = "/metrics";
+      description = lib.mdDoc "Metrics expose path";
+    };
+  };
+  serviceOpts = {
+    serviceConfig = {
+      DynamicUser = false;
```
Comment (NickCao) [inline nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix:57]: There is no need to explicitly set `DynamicUser = false`, or you mean true?

File: nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      RuntimeDirectory = "prometheus-mongodb-exporter";
+      ExecStart = ''
+        ${pkgs.prometheus-mongodb-exporter}/bin/mongodb_exporter \
+          --mongodb.uri=${cfg.uri}
+          ${if cfg.collectAll then "--collect.all" else ""} \
```
Comment (NickCao) [inline nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix:62]:
```suggestion
          ${lib.optionalString cfg.collectAll then "--collect.all"} \
```

File: nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+          --collector.collstats=${concatStringsSep "," cfg.collStats} \
+          --collector.indexstats=${concatStringsSep "," cfg.indexStats} \
+          --web.listen-address=${cfg.listen} \
+          --web.telemetry-path=${cfg.telemetryPath} \
+          ${concatStringsSep " " cfg.extraFlags}
```
Comment (NickCao) [inline nixos/modules/services/monitoring/prometheus/exporters/mongodb.nix:67]:
```suggestion
          ${lib.escapeShellArgs cfg.extraFlags}
```

**Review by NickCao — COMMENTED** [review]
Please split the package addition and the module addition to two commits.

The commit messages should look like:
- `prometheus-mongodb-exporter: init at <version>`
- `nixos/prometheus-mongodb-exporter: init module`

Otherwise LGTM.

**Comment by undefined-moe** [comment]
fixed.

**Comment by NickCao** [comment]
@ofborg build prometheus-exporter-mongodb

**Pull request merged**